Add support for custom attributes in Table subclasses

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

Quoting the docs:

One simple customization that can be useful is adding new attributes to the table object.  There is nothing preventing setting an attribute on an existing table object, for example ``t.foo = 'hello'``.  However, this attribute would be ephemeral because it will be lost if the table is sliced, copied, or pickled. Instead, you can add persistent attributes as shown in this example.
```
  from astropy.table import Table, TableAttribute

  class MyTable(Table):
      foo = TableAttribute()
      bar = TableAttribute(default=[])
      baz = TableAttribute(default=1)

  t = MyTable([[1, 2]], foo='foo')
  t.bar.append(2.0)
  t.baz = 'baz'
```
Some key points:

- A custom attribute can be set when the table is created or using the usual syntax for setting an object attribute.
- A custom attribute always has a default value, either explicitly set  in the class definition or ``None``.
- The attribute values are stored in the table ``meta`` dictionary. This is  the mechanism by which they are persistent through copy, slice, and  serialization such as pickling or writing to an ECSV ASCII file.

### Background

I used a version of this extensively in one Chandra operations package (including subclassing the attribute descriptor for even more customization).  After deciding to use the same thing in another package, I thought it might be more widely useful.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

pllim commented:
Oh, is this like, for no better words to put it, FITS header keywords?

Failure is related:
```
ImportError: `import yaml` failed, PyYAML package is required for serializing mixin columns
```

mhvk commented on astropy/table/table.py:
```diff
@@ -3674,3 +3683,42 @@ def __setstate__(self, state):
+    def __get__(self, instance, owner):
+        try:
+            return instance.meta[self.name]
+        except AttributeError:
```
This would also go wrong if the `TableAttribute` was attached to a class without a `meta`, leading to a difficult-to-debug state. I'd just do 
```
if instance is None:
    return self
```
which is very fast (and keeps the more relevant `KeyError` very close). Or, if you prefer the error route,
```
if instance is None:
    return self
else:
    raise
```

mhvk commented on astropy/table/table.py:
```diff
@@ -3674,3 +3683,42 @@ def __setstate__(self, state):
+        reserved = ('data', 'masked', 'names', 'dtype',
+                    'meta', 'copy', 'rows', 'copy_indices',
+                    'units', 'descriptions')
+        if hasattr(Table, name) or name in reserved:
```
Shouldn't this be `hasattr(owner, name)`? If the subclass defines a new method, presumably that shouldn't be overwritten either.

More because I'm a bit confused: the `reserved` set look like the keywords in `__init__` - maybe get it automatically by inspection? Or is it in fact needed? (`meta` and `copy` would not seem needed, since they exist already).

taldcroft replied:
Yes, this should be `owner` not hard-coded Table.  I'm thinking now about `reserved` as a class attribute of `MetaAttribute`, so users can subclass.  Then perhaps we still end up with `TableAttribute` as a subclass of `MetaAttribute`.

The idea of `reserved` is to explicitly avoid collisions in the class `__init__` that might cause confusion.  Especially if this is put out there as something to use on any class, then there is really no way to *reliably* infer the API even with inspect for classes that hav a catch-all `**kwargs`.

mhvk replied:
This perhaps is an argument to stick with `TableAttribute` here for now, and give ourselves the flexibility to implement a more general `MetaAttribute` later, from which `TableAttribute` can then inherit.

taldcroft commented:
@mhvk - about the what's in store, I can show you the original use case.  For Chandra operations when we go to a new attitude for observing, the star tracker needs a catalog of up to 8 stars to track.  So below is a class that inherits from `Table` to store that catalog of stars along with a whole bunch of related meta data about the observation.  I quickly got tired of `starcat.meta['something']` instead of `starcat.something` and also realized that defaults and some other customization would help alot.

https://github.com/sot/proseco/blob/78611dd7a0d5573f97b4753b4dfaf44ebc1f944f/proseco/core.py#L515

It's funny that you suggested `MetaAttribute`, that is exactly the original name I used.  And my first draft of this PR actually put this into `utils/metadata.py` as a somewhat generic thing.  But then I couldn't think of a different generic use case and thought for users that `TableAttribute` might be more obvious since "meta" is so overloaded with different meanings.

p.s. I might have made a mistake early on inheriting from Table instead of making a container class, but there you go.

taldcroft commented:
@mhvk - thanks for the review and good thoughts.  I had not considered the immediate potential application to `Column` but that could work.  I suppose `NDData` as well?  I think I can incorporate most of what you said.

About using this now in core, there is always the question of performance.  As you add layers of abstraction the code might look nicer but is also almost certain to run more slowly.  But anyway, that would be a discussion for another PR.

taldcroft commented:
One other question I had is if this should bundle these attributes in a sub-dict within `meta`, e.g. `meta['__attributes__'] = {}`.  Slows things down, but prevents collisions, and of course `__attributes__` could be configurable as a `MetaAttribute` class attribute.

mhvk commented:
Both ideas sound good, though as usual there may be pros and cons. Storing the attributes in `meta['__attributes__']` has a nice analogy to things like `__serialized_columns__` in that it is clearer that it is "something else" that is carried here, but on the other hand that perhaps removes the connection with an easy short-cut to something stored in `meta`.

Allowing the place where the attributes get stored to be configurable seems definitely a good idea, and I like how it allows easy extension to `.info.meta`. But it is a complication that is not yet needed, and I don't think anything in what you have would prevent one from later adding it.

Anyway, overall I think this is a nice little addition, and I can see use cases also internal to astropy - say, for the `IERS` tables where we now have a couple of ad-hoc attributes that might as well get saved (if one ever saved a table, that is...).

taldcroft commented on astropy/utils/metadata.py:
```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
+
+    def __set_name__(self, owner, name):
+        import inspect
+        params = [param.name for param in inspect.signature(owner).parameters.values()
```
In the end I decided to just go with `inspect` since that works for the current use case.  If we end up needing more flexibility then that can be added later with no API change here.

mhvk replied:
Nice.

taldcroft commented:
@mhvk - I think I addressed your comments and suggestions.  In particular there is now a `MetaAttributes` base class and a (trivial) `TableAttributes` subclass . The reason for the existence of the `TableAttributes` version is to have table-specific documentation and make it easier for users as a part of the `table` subpackage.  I added a lot more documentation in the doc strings aimed at developers.

mhvk commented on astropy/utils/metadata.py:
```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
+            return self
+
+        # Get the __attributes__ dict and create if not there already.
+        meta = instance.meta.setdefault('__attributes__', {})
```
Maybe call the variable `attributes`? I was confused reading this, as I wondered whether `instance.meta` was returned - I had never realized `setdefault` returns anything!

If so, also for `__set__`

taldcroft replied:
Done.

mhvk commented on astropy/utils/metadata.py:
```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
+        # Get the __attributes__ dict and create if not there already.
+        meta = instance.meta.setdefault('__attributes__', {})
+        meta[self.name] = value
+
```
Maybe define a `__del__` method while we are at it?

taldcroft replied:
I started down that path but realized that for this current implementation `__delete__` doesn't quite make sense.  You cannot actually delete a `MetaAttribute` because it always comes back with the default.  So you would get:
```
>>> del obj.meta_attr
>>> obj.meta_attr
# prints the default for meta_attr
```
Now we *could* change `__init__` so that there is a difference between a meta attribute that has an explicitly specified default and one that does not:
```
class MyClass:
    ma1 = MetaAttribute()
    ma2 = MetaAttribute(default=5)

obj = MyClass()
obj.ma2  # returns 5
obj.ma1  # AttributeError (current code returns None)
```
I'm not sure which way to go here.  In my real-world applications where this was developed, it was convenient to have that default of `None`.  OTOH it is a bit more Pythonic that accessing an attribute that has not been set should raise AttributeError unless a default is explicitly provided.  If we do this then `__delete__` would make sense.

taldcroft replied:
I started writing this implementation to see how it looks.  But I got really stuck at how confusing it seems in practice to have explicitly defined `ma1` as a class attribute and then immediately afterword get an `AttributeError: 'MyClass' has no attribute 'ma1'` when it is right there as a class attribute.

I'm kind of leaning back to these always having an implicit default of None, and thinking of them as a bit more like properties than attributes.  Maybe they should be called `MetaProperty`?

taldcroft replied:
And in fact when I started this whole thing for my own work, I actually wrote the first couple as properties and then realized that I could generalize and reduce boilerplate code by making the descriptor.

taldcroft replied:
For the record, this is what I started doing.
```
--- a/astropy/utils/metadata.py
+++ b/astropy/utils/metadata.py
@@ -446,8 +446,12 @@ class MetaAttribute:
     :param default: default value
 
     """
-    def __init__(self, default=None):
-        self.default = default
+    def __init__(self, **kwargs):
+        if 'default' in kwargs:
+            self.default = kwargs.pop('default')
+        if kwargs:
+            raise TypeError(f"__init__() got unexpected keyword argument "
+                            f"'{list(kwargs)[0]}'")
 
     def __get__(self, instance, owner):
         # When called without an instance, return self to allow access
@@ -460,10 +464,11 @@ class MetaAttribute:
         try:
             value = attributes[self.name]
         except KeyError:
-            if self.default is not None:
-                attributes[self.name] = deepcopy(self.default)
-            # Return either specified default or None
-            value = attributes.get(self.name)
+            if hasattr(self, 'default'):
+                value = attributes[self.name] = deepcopy(self.default)
+            else:
+                raise AttributeError(f"'{instance.__class__.__name__}' object has no "
+                                     f"attribute 'self.name'")
         return value
 
     def __set__(self, instance, value):
```

mhvk commented on astropy/utils/metadata.py:
```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
+        try:
+            value = meta[self.name]
+        except KeyError:
+            if self.default is not None:
```
I wondered whether we should return `deepcopy(self.default)` here independent of the value of `default`, i.e., never set the attribute? But now I realize that if `self.default = {}`, that would go rather badly wrong... (that example of course also is a good reason for the copy).  So, :+1: to the implementation, but perhaps add a comment?

bsipocz commented:
I'm not sure what's going on with the CI here, please try to rebase and address and reject the review comments so we can merge.

mhvk commented:
@taldcroft - sorry I sent you on such a wild-goose chase! The additional complexity indeed seems beyond what it was worth to get an `AttributeError` - if we ever find we need it after all, it may still be possible to do it with a special value for `no default` that is not `None`. Anyway, happy to see this in!
